### PR TITLE
Fix issues building Mantid with Poco 1.6.* (version 2)

### LINKED ISSUE #11815: Poco 1.6.0 upgrade causing 5 tests to fail
This issue was originally [TRAC 10976](http://trac.mantidproject.org/mantid/ticket/10976)

This ticket is blocked by : 
 - [X] [TRAC11117](http://trac.mantidproject.org/mantid/ticket/11117)

Today homebrew upgraded poco to 1.6.0. After upgrading the following tests fail. 

 81 - KernelTest_StringsTest (Failed)
125 - GeometryTest_FitParameterTest (Failed)
476 - DataHandlingTest_GroupDetectors2Test (Failed)
525 - DataHandlingTest_LoadRKHTest (Failed)
851 - CurveFittingTest_FitMWTest (Failed)

### PR DESCRIPTION
**Pull request by quantumsteve**
This fixes #11815 and is a continuation of PR #15150. I created a new branch to fix a bunch of files I accidentally clang-formatted while solving merge conflicts. The message from the previous PR is repeated below.

---

The behavior of `Poco::StringTokenizer` changed between versions 1.4 and 1.6. This wraps the Poco::StringTokenizer class with [a small function](http://tcbrindle.github.io/posts/a-quicker-study-on-tokenising/) using the STL. I Initially tried to wrap the StringTokenizer from Poco 1.6.1, but found maintaining the 1.4 behavior required a slow and ugly implementation. I added unit tests to check that this tokenizer consistently parses various strings used in other algorithms. 
### breaking changes:
1. Poco 1.4 doesn't include the last token if it is empty, while 1.6 does. I've added an option to give either behavior.
2. Poco 1.4 and 1.6  handle struggle to tokenize strings with whitespace. I tried to follow what would occur if one were to use boost::tokenizer, then potentially trim and/or remove empty tokens.
   1. Poco 1.4 fails to find two tokens in str="  ", sep=" " and options =`StringTokenizer::TOK_TRIM`
   2. Poco 1.6 fails to find two tokens in str="2- ", sep="-" and and options =`StringTokenizer::TOK_TRIM`

I added three performance tests tokenizing a large string with various options. I hope it'll give us a baseline from which to consider optimizations with `boost::string_ref` or an alternative implementation using `boost::spirit:qi`. Any suggestions for minimizing cache effects in the performance tests? I'm concerned the typical tricks of initializing a large array of random values would interfere with the automatic generation of times.

Documentation: I'll add Poco 1.6.\* support to the 3.7 release notes once the template is available.

Code review: This PR changes the StringTokenizer's behavior, eliminating many quirks of the `Poco::StringTokenizer` implementation. I fixed changes caused by whitespace and empty tokens caught by the unit tests, but would appreciate another set of eyes. 

Testing: Please check if I missed unit testing important cases. Does the current behavior regarding whitespace and empty tokens now seem reasonable?

### CODE REVIEW

File: Framework/DataHandling/inc/MantidDataHandling/GroupDetectors2.h [hunk 1]
```diff
@@ -239,8 +241,8 @@ class DLLExport GroupDetectors2 : public API::Algorithm {
     /// spectrum number to the this
     EMPTY_LINE = 1001 - INT_MAX, ///< when reading from the input file this
     /// value means that we found any empty line
-    IGNORE_SPACES = Poco::StringTokenizer::TOK_TRIM ///< equal to
-    /// Poco::StringTokenizer::TOK_TRIM but
+    IGNORE_SPACES = Mantid::Kernel::StringTokenizer::TOK_TRIM ///< equal to
+    /// Mantid::Kernel::StringTokenizer::TOK_TRIM but
     /// saves some typing
   };
```
Comment (MichaelWedel) [inline Framework/DataHandling/inc/MantidDataHandling/GroupDetectors2.h]: I did not spot this before, but looking closer it seems these two instances of `IGNORE_SPACES` now mean different things, while they meant the same before.

Was this uncovered by a test?

Comment (MichaelWedel) [inline Framework/DataHandling/inc/MantidDataHandling/GroupDetectors2.h]: Modified the code for understanding the first part, this was caught by failing tests - all good.

**Comment by MichaelWedel** [comment]
Just to clarify, was the clang-format intentional and supposed to go in? I'm happy to merge it, just want to make sure this is what you meant to happen.

While going through the test cases again I found two cases that are not tested:

1. Multiple separators behave as expected.
2. Empty separator string simply returns one string.

Wrote and tried these two small tests:
```
  void test_StringTokenizer_multipleSeparators() {
    auto tokenizer = Mantid::Kernel::StringTokenizer("1,2,3,-5,6", ",-");
    std::vector<std::string> expected{"1", "2", "3", "", "5", "6"};
    TS_ASSERT_EQUALS(tokenizer.asVector(), expected);
  }

  void test_StringTokenizer_emptySeparators() {
    auto tokenizer = Mantid::Kernel::StringTokenizer("1,2,3,-5,6", "");
    std::vector<std::string> expected{"1,2,3,-5,6"};
    TS_ASSERT_EQUALS(tokenizer.asVector(), expected);
  }
```
Both work as expected - I'd say it's probably still useful to add them as test cases. If nothing else they at least capture current behavior which makes it easier if this has to be replaced again in some distant future.

I tried treating the second case with an additional if-statement to return early but that did not change speed at all (modified the long string performance test).

Otherwise the tests pass and I tried modifying the cases that were unclear to me and they were caught by failing tests.

**Comment by quantumsteve** [comment]
@MichaelWedel No, the clang-format changes were a mistake. Clang-format and merge conflicts don't mix well. I checked out all MantidQt/* and MantidPlot/* from master and then replaced the two files I actually touched in those directories.

File: Framework/DataHandling/inc/MantidDataHandling/GroupDetectors2.h (hunk 1 above)
Comment (quantumsteve) [inline Framework/DataHandling/inc/MantidDataHandling/GroupDetectors2.h]: Glad to see potential issues here were caught by failing tests.

**Comment by quantumsteve** [comment]
I added the test cases you suggested above. Are we ready to merge?

**Comment by MichaelWedel** [comment]
A part of the code review was done on the original pull request, this PR addressed all points that were raised there. I modified some of the code and ran some of the unit tests, from my point of view this is ready to merge. :shipit:

**Pull request merged**